src/smt/theory_str_noodler/theory_str_noodler.cpp

```diff
@@ -489,27 +461,14 @@ namespace smt::noodler {
         expr *e = ctx.bool_var2expr(v);
         expr *e1 = nullptr, *e2 = nullptr;
         if (m_util_s.str.is_prefix(e, e1, e2)) {
-            // INFO done in relevant_eh, because there was some problem with adding axioms, Z3 added wrong axiom if it was done here and not in relevant_eh for some reason
-            // INFO in seq_theory they do similar stuff, and they do it here and not in relevant_eh, they also use skolem functions to handle it, might be worth to investigate
-
-            // if (is_true) {
-            //     handle_prefix(e);
-            // } else {
-            //     util::throw_error("unsupported predicate");
-            //     //handle_not_prefix(e);
-            // }
+            // already handled in relevant_eh. It suffices to handle is_prefix only once as it is fully 
+            // axiomatized using basic string constraints ((dis)equations, regexes, and lengths)
         } else if (m_util_s.str.is_suffix(e, e1, e2)) {
-            // INFO done in relevant_eh, because there was some problem with adding axioms, Z3 added wrong axiom if it was done here and not in relevant_eh for some reason
-            // INFO in seq_theory they do similar stuff, and they do it here and not in relevant_eh, they also use skolem functions to handle it, might be worth to investigate
-
-            // if (is_true) {
-            //     handle_suffix(e);
-            // } else {
-            //     util::throw_error("unsupported predicate");
-            //     // handle_not_suffix(e);
-            // }
+            // already handled in relevant_eh. It suffices to handle is_suffix only once as it is fully 
+            // axiomatized using basic string constraints ((dis)equations, regexes, and lengths)
         } else if (m_util_s.str.is_contains(e, e1, e2)) {
-            // FIXME could the problem from previous two also occur here? if yes, handle_contains and handle_not_contains should be in relevant_eh BUT m_not_contains_todo should be updated probably here (if applicable), so we can return unknown
+            // notcontains cannot be fully axiomatized. We need to add it among string constraints solved later in final_check 
+            // (it is not sufficient to add it only once at the beginning of the solver run as it max occurr in different SAT assignments).
```

Comment:
For me this still does not explain why there is no problem with handling `contains` here, but there is problem for the previous ones. Generally speking, I think we should move towards handling everything here and not in `relevant` as we do now, or at least Z3 does it this way.

Reply:
I think it could be safe to move the handling of `contains` to the relevant only (It needs to be checked on the benchmark, I haven't tried, so I left it untouched). In general it would be nice to have it in `assign_eh` but it requires storing information about which predicates were already axiomatized at which decision levels (the `scoped_vector` datastructure is probably fine for that but you need to somehow properly integrate to the noodler's internals)..

Reply:
But we do not do the storing about this information currently for `contains` right? So if we keep it here for now, it could lead to bugs, or am I wrong?

Reply:
Currently the `handle_contains` is redundant there because it is handled already in `relevant_eh` (since we cache the occurrences, it does nothing in `assign_eh`). I removed it and it works ok (no wrong results nor performance degradation).